src/combineReducers.js

```diff
@@ -74,6 +74,46 @@ function assertReducerSanity(reducers) {
   })
 }
 
+function getReducers(config) {
+  var reducers = {}
+
+  Object.keys(config).map(key => {
+    if (typeof config[key] === 'function') {
+      reducers[key] = config[key]
+    }
+  })
+
+  return reducers
+}
+
+function checkStateShape(keys, reducers, state, action) {
+  var hasChanged = false
+  var nextState = {}
+  for (var i = 0; i < keys.length; i++) {
+    var key = keys[i]
+    var reducer = reducers[key]
+    var previousStateForKey = state[key]
+    var nextStateForKey = reducer(previousStateForKey, action)
+
+    if (typeof nextStateForKey === 'undefined') {
+      var errorMessage = getUndefinedStateErrorMessage(key, action)
+      throw new Error(errorMessage)
+    }
+
+    nextState[key] = nextStateForKey
+    hasChanged = hasChanged || nextStateForKey !== previousStateForKey
+  }
+
+  return hasChanged ? nextState : state
+}
+
+function showWarningMessage(state, reducers, action) {
```

Comment:
This should be called `checkStateShape` because this is what it does. `showWarningMessage` is misleading because it feels like a warning would be called every time, whereas in practice it is only called if the state shape is unexpected.